sql/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java

```diff
@@ -217,18 +216,22 @@ private static void ensureUnionOutputsHaveCompatibleTypes(QueriedRelation left,
         for (int i = 0; i < leftOutputs.size(); i++) {
             Symbol leftSymbol = leftOutputs.get(i);
             Symbol rightSymbol = rightOutputs.get(i);
-            DataType leftDataType = leftSymbol.valueType();
-            DataType rightDataType = rightSymbol.valueType();
-
-            if (leftDataType.equals(rightDataType)) {
-                continue;
-            } else if (rightDataType.precedes(leftDataType) &&
-                       leftSymbol.valueType().isConvertableTo(rightSymbol.valueType())) {
-                left.outputs().set(i, leftSymbol.cast(rightDataType));
-            } else if (rightSymbol.valueType().isConvertableTo(leftSymbol.valueType())) {
-                rightOutputs.set(i, rightSymbol.cast(leftDataType));
-            } else {
-                throw new UnsupportedOperationException("Corresponding output columns must be compatible for all parts of a UNION");
+            Tuple<Byte, Symbol> result = Symbols.tryUpcastToCompatibleType(leftSymbol, rightSymbol);
```

Comment:
Instead of the tuple, how about the following:

```java
leftSymbol = tryUpcast(leftSymbol, rightSymbolType);
rightSymbol = tryUpcastAndFailOtherwise(rightSymbol, leftSymbolType);
```

This will cast one of the two and throw an error if upcasting is not possible.